### ISSUE TITLE
Auto-generated test failure using <link.h> on Debian 9.

### ISSUE THREAD

**Issue by vext01** [rust-lang-nursery/rust-bindgen#1370]
### Input C/C++ Header

Sorry, I know you asked for no `#include`s, but I'm at a loss as to what the predicate script should do when I'm just importing types from the C library -- If you can't reproduce from the info below, then could you help me with the predicate script plz? :)

```C++
#define _GNU_SOURCE
#include <link.h>
```

### Bindgen Invocation

```Rust
    let bindings = bindgen::Builder::default()
                           .header(WRAPPER_HEADER)
                           .generate()
                           .expect("bindgen failed");
```

### Actual Results

When included, like this:
```Rust
#![allow(non_upper_case_globals)]
#![allow(non_camel_case_types)]
#![allow(non_snake_case)]
#![allow(dead_code)]
include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
```

```
$ RUST_BACKTRACE=1 cargo test bindgen_test_layout_La_x86_64_retval
---- bindgen_test_layout_La_x86_64_retval stdout ----
thread 'bindgen_test_layout_La_x86_64_retval' panicked at 'assertion failed: `(left == right)`
  left: `224`,
 right: `240`: Size of: La_x86_64_retval', /home/vext01/research/bg/target/debug/build/bg-29903e2596eaef66/out/bindings.rs:3:221231
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:221
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:475
   5: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:390
   6: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:345
   7: bg::bindgen_test_layout_La_x86_64_retval
             at ./target/debug/build/bg-29903e2596eaef66/out/bindings.rs:3
   8: bg::__test::TESTS::{{closure}}
             at ./target/debug/build/bg-29903e2596eaef66/out/bindings.rs:3
   9: core::ops::function::FnOnce::call_once
             at /checkout/src/libcore/ops/function.rs:223
  10: <F as alloc::boxed::FnBox<A>>::call_box
             at libtest/lib.rs:1459
             at /checkout/src/libcore/ops/function.rs:223
             at /checkout/src/liballoc/boxed.rs:642
  11: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:102
```

This is the only failing test.

### Expected Results

Test passes.

This is on Debian 9 x86_64, using today's Rust nightly.

FWIW, OpenBSD's `<link.h>` does work on an OpenBSD system.

Any ideas?

**Comment by emilio** [comment]
Is there any chance of knowing the preprocessed input? Otherwise it's going to be hard to figure out.

See: https://github.com/rust-lang-nursery/rust-bindgen/blob/master/CONTRIBUTING.md#isolating-your-test-case

If you feel like trying creduce to reduce it it'd be awesome as well!

**Comment by emilio** [comment]
> Sorry, I know you asked for no #includes, but I'm at a loss as to what the predicate script should do when I'm just importing types from the C library -- If you can't reproduce from the info below, then could you help me with the predicate script plz? :)

Oh, the predicate script is only for reducing it. The `.i` file would still be really helpful to repro it locally.

**Comment by vext01** [comment]
Gzipped `.i` file attached. Hopefully that helps.

[__bindgen.i.gz](https://github.com/rust-lang-nursery/rust-bindgen/files/2309589/__bindgen.i.gz)

**Comment by vext01** [comment]
Any joy?

Thanks

**Comment by emilio** [comment]
Sorry, I haven't had the chance to look at this in detail yet.

**Comment by vext01** [comment]
Hi emilio,

How does one go about debugging this test failure? I see the test in bindings.rs, but it's large and all on one line. Is there a way to get this in a more human-readable format?

We have a PR blocked on this issue, so I'm keen to push on.

Thanks

**Comment by emilio** [comment]
If you have rustfmt installed you should get the bindings properly formatted.

**Comment by vext01** [comment]
> I think bindgen is not dealing with the conditionally compiled fields properly.

Actually, that statement can't be right.

Any ideas?

**Comment by vext01** [comment]
Here's the real problem:

```
long double lrv_st0;
```

becomes:
```
pub lrv_st0: f64,
```

Which I think is incorrect.

On my system, a long double is 16 bytes long, whereas an f64 is 8, and this messes up the size of the struct.

Related: https://github.com/rust-lang-nursery/rust-bindgen/issues/550

**Comment by vext01** [comment]
Hey Emilio,

Sounds good, but why conditionally?

Cheers

**Comment by emilio** [comment]
Because we want to support old rust versions, so it'd need to be behind a `rust_target` check.

**Comment by emilio** [comment]
I took a bit of spare time to implement the first step to properly do that: https://github.com/rust-lang-nursery/rust-bindgen/pull/1391. That fixes integer generation. Then I need to teach the rest of the codegen to conditionally support that, but shouldn't be terrible.

**Comment by emilio** [comment]
I ended up doing the whole thing :)

**Comment by emilio** [comment]
Note that there's an unexpected gotcha here: https://github.com/rust-lang/rust/issues/54341.

**Comment by emilio** [comment]
Have you enabled `rust_target(Rust_Target::Stable_1_26)`? Otherwise behavior should be unchanged.

**Comment by emilio** [comment]
> Ideally, Rust (or the libc crate) would have a long_double type which aligns properly. How likely is it that we could convince them?

I think I did file an issue on this long time ago... But I can't find it now. It's worth a shot probably :)

**Comment by vext01** [comment]
> Have you enabled rust_target(Rust_Target::Stable_1_26)? 

Um, no. I thought nightly would be enough. Let me give it a go.

**Comment by vext01** [comment]
Right, so this makes a difference:
```
---- p_ffi::bindgen_test_layout_La_x86_64_retval stdout ----
thread 'p_ffi::bindgen_test_layout_La_x86_64_retval' panicked at 'assertion failed: `(left == right)`
  left: `8`,
 right: `16`: Alignment of La_x86_64_retval', /home/vext01/research/phdrs_pv/target/debug/build/phdrs-bbf2d34987b05535/out/bindings.rs:7273:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- p_ffi::bindgen_test_layout_La_x86_64_vector stdout ----
thread 'p_ffi::bindgen_test_layout_La_x86_64_vector' panicked at 'assertion failed: `(left == right)`
  left: `8`,
 right: `16`: Alignment of La_x86_64_vector', /home/vext01/research/phdrs_pv/target/debug/build/phdrs-bbf2d34987b05535/out/bindings.rs:7077:5
```

**Comment by emilio** [comment]
I pushed a commit that should fix those. It's a bit of a work-around, but it only improves correctness so I think we should land it.

**Comment by emilio** [comment]
And rustc already warns when using `u128` as a raw type via FFI, so I think that should close this.

> warning: `extern` block uses type `u128` which is not FFI-safe: 128-bit integers don't currently have a known stable ABI

**Comment by vext01** [comment]
Hi emilio,

So should my tests pass now? Using the latest version of your u128 branch,  I still have one test failure:
```
---- p_ffi::bindgen_test_layout_La_x86_64_vector stdout ----
thread 'p_ffi::bindgen_test_layout_La_x86_64_vector' panicked at 'assertion failed: `(left == right)`
  left: `8`,
 right: `16`: Alignment of La_x86_64_vector', /home/vext01/research/phdrs_pv/target/debug/build/phdrs-766b5f3742f78b0d/out/bindings.rs:7077:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by emilio** [comment]
Oh, interesting, that is not a struct but a union with alignment attributes. Mind filing a new issue for that one? I'm not on my laptop now but I think it should be pretty easy to create a reduced test-case.

**Comment by vext01** [comment]
Sure, I'll raise another issue.

**Comment by vext01** [comment]
Hi emilio,

Quick question. For how long will I need the `.rust_target(bindgen::RustTarget::Stable_1_26)` hack in `build.rs`?

Thanks

**Comment by emilio** [comment]
It's not really a hack, it's there so bindgen will generate repr(align) attributes. We could deprecate support for older rust versions I guess, and make one version the default from now on, but it's not clear what version should that be.

**Comment by vext01** [comment]
Maybe this is not possible, but wouldn't you just want something which works with whatever version of rust the user is currently running?
